File: hw/sanchez/u3/hw18OOPConceptsZoo/OOPConceptsZoo/src/main/java/ec/edu/espe/oopconceptszoo/controller/DatabaseConnection.java
package ec.edu.espe.oopconceptszoo.controller;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;


/**
 *
 * @author Joel Sanchez, The Softwarriors, @ESPE
 */
public class DatabaseConnection {
    private static DatabaseConnection instance;
    private MongoClient mongoClient;
    private MongoDatabase database;
    
    private static final String CONNECTION_STRING = "mongodb+srv://Joel:Joel@cluster0.aex8od4.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
    private static final String DATABASE_NAME = "OOPConceptsZoo";
    
    private DatabaseConnection() {
        try {
            mongoClient = MongoClients.create(CONNECTION_STRING);
            database = mongoClient.getDatabase(DATABASE_NAME);
            System.out.println("Connected to MongoDB Atlas");
        } catch (Exception e) {
            System.err.println("Error connecting to MongoDB Atlas: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    public static DatabaseConnection getInstance() {
        if (instance == null) {
            instance = new DatabaseConnection();
        }
        return instance;
    }
    
    public MongoCollection<Document> getCollection(String collectionName) {
        return database.getCollection(collectionName);
    }
    
    public void close() {
        if (mongoClient != null) {
            mongoClient.close();
        }
    }
    
    public boolean isConnected() {
        try {
            database.listCollectionNames().first();
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}

File: hw/sanchez/u3/hw18OOPConceptsZoo/ZooPython/controller/database_connection.py
import os
from pymongo import MongoClient
from pymongo.collection import Collection
from dotenv import load_dotenv

load_dotenv()

class DatabaseConnection:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialize()
        return cls._instance
    
    def _initialize(self):
        self.connection_string = os.getenv('MONGODB_URI')
        self.database_name = os.getenv('DATABASE_NAME', 'OOPConceptsZoo')
        try:
            self.client = MongoClient(self.connection_string)
            self.database = self.client[self.database_name]
            print("Connected to MongoDB Atlas")
        except Exception as e:
            print(f"Error connecting to MongoDB Atlas: {e}")
    
    def get_collection(self, collection_name: str) -> Collection:
        return self.database[collection_name]
    
    def is_connected(self) -> bool:
        try:
            self.database.list_collection_names()
            return True
        except Exception:
            return False
    
    def close(self):
        if self.client:
            self.client.close()